clientapi/routing/register_test.go

```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
+			*defaults,
+			// len 600
+			"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("'password' >%d characters", defaults.MaxPasswordLength)),
+			},
+		},
+		{
+			"default allow long enough",
+			*defaults,
+			"thisisalongenoughpassword",
+			nil,
+		},
+		{
+			"default allow with symbols",
+			*defaults,
+			"ih@ve$ome$ymbol$_here",
+			nil,
+		},
+		{
+			"set min reject too short",
+			*custom,
+			"abcd",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: min %d chars", custom.MinPasswordLength)),
+			},
+		},
+		{
+			"set max reject too long",
+			*custom,
+			// len 33
+			"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("'password' >%d characters", custom.MaxPasswordLength)),
+			},
+		},
+		{
+			"set symbols not enough",
+			*custom,
+			"thi$i$apasswordshouldbelong",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: minimum %d symbols", custom.MinNumberSymbols)),
+			},
+		},
+		{
+			"require mixed case but none given",
+			*custom,
+			"haha_all_lowercase_cant_catch_me",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword("password must have uppercase and lowercase letters"),
+			},
+		},
+		{
+			"custom settings allow",
+			*custom,
+			"$0me_$up3r_$trong_P@ass",
+			nil,
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			response := validatePassword(test.password, test.config)
+			if test.expected == nil && response != nil {
+				t.Errorf("expected password to be validated. got error: %s", response.JSON.(*jsonerror.MatrixError).Err)
+			} else if test.expected != nil && response == nil {
+				t.Errorf("expected password to fail, but was validated")
+			} else if test.expected != nil && test.expected.Code != response.Code {
+				t.Errorf("expected error code %d, got %d", test.expected.Code, response.Code)
+			} else if test.expected != nil && response != nil {
+				matrixError := response.JSON.(*jsonerror.MatrixError)
+				expectedError := test.expected.JSON.(*jsonerror.MatrixError)
+
+				if expectedError.Err != matrixError.Err {
```

Comment:
Please don't do this. Asserting error strings is really a pain, we have error codes for a reason, use that instead.